### PR TITLE
Support credit card payments through Adyen's API

### PR DESCRIPTION
**Pull request by luukveenis**
Currently we only support payments that go through Adyen's Hosted Payment Pages (HPP). This change adds a new payment method for credit cards and everything required to support processing credit card payments directly through Adyen's API, using their client side encryption (CSE) library.

Credit card payments through Adyen work similar to HPP payments in that it's an asynchronous API. We make a payment request, Adyen tells us whether it was received successfully or not, and at some point in the future we get a notification if it succeeded or not. Most of the handling of notifications uses the existing HPP code with minimal modifications. We can capture, cancel, refund, and partially refund payments through Solidus and all operations will be handled like HPP, where the payment state is put into `processing` on a request and transitioned after we receive a notification from Adyen.

One of the challenges with Adyen is that they don't support typical payment profiles provided by services like Stripe or Braintree. Instead, Adyen supports "recurring payment contracts". These still store the user's card details with Adyen and allow them to be re-used, but they require an initial payment request to be created (such as an authorization). Because of this, we have to do the authorization during the `payment` checkout step when we still have access to the encrypted credit card data. That means by the time we get to `confirm` we've already authorized the payment and either fake the authorization to advance the payment or simply `capture` at that point, depending if the payment method has auto-capture enabled or not.

The `payment_profiles_supported?` method returns false because Solidus assumes you're following the typical flow when this is true, which causes a number of issues. However, the recurring contract is still created and the reference is stored in the `gateway_customer_profile_id` attribute on the source so that we can re-use the details to create payments or make modifications later.

The biggest issue with authorizing during the payment step is that we may be left with outstanding authorizations if the user abandons checkout on the confirm step. This is something to be aware of and anyone using credit card payments should periodically cancel these open authorizations.

### CODE REVIEW

**Comment by cbrunsdon** [comment]
I worked closely with luuk (and before him martin) and am :+1: on merging this. It should be completely inconsequential to anyone not using CC's.

Some bits of it are pretty unfortunate, but what part of trying to connect adyen to solidus isn't.

**Comment by Dkendal** [comment]
requires a rebase

File: app/controllers/spree/adyen_notifications_controller.rb [hunk 1]
```diff
@@ -29,5 +30,12 @@ def authenticate
     def accept
       render text: "[accepted]"
     end
+
+    def notification_exists? params
+      AdyenNotification.where(
```
Comment (Dkendal) [inline app/controllers/spree/adyen_notifications_controller.rb]: `.exists?` would also be appropriate.

File: solidus-adyen.gemspec [hunk 2]
```diff
@@ -19,14 +19,12 @@ Gem::Specification.new do |spec|
   spec.require_paths = ["lib"]
 
   spec.add_runtime_dependency "adyen", "~> 2.2.0"
-  spec.add_runtime_dependency "solidus_core", "~> 1.1"
+  spec.add_runtime_dependency "solidus_core"
```
Comment (Dkendal) [inline solidus-adyen.gemspec]: any reason for unlocking the dep?

**Comment by Dkendal** [comment]
thanks guys lgtm

**Comment by oleksiidrobiazko** [comment]
@luukveenis and other guys, thank you, great work.
Could you tell me please it's not released? In master but not released, right?

**Comment by Dkendal** [comment]
@oleksiidrobiazko you are correct, this is currently an unreleased change in master

**Pull request merged**